allennlp/data/dataset_readers/semantic_parsing/grammar_based_text2sql.py

```diff
@@ -0,0 +1,140 @@
+        We use this to define the input representation for the text.  See :class:`TokenIndexer`.
+        Note that the `output` tags will always correspond to single token IDs based on how they
+        are pre-tokenised in the data file.
+    cross_validation_split_to_exclude : ``int``, optional (default = None)
+        Some of the text2sql datasets are very small, so you may need to do cross validation.
+        Here, you can specify a integer corresponding to a split_{int}.json file not to include
+        in the training set.
+    """
+    def __init__(self,
+                 schema_path: str,
+                 use_all_sql: bool = False,
+                 remove_unneeded_aliases: bool = True,
+                 token_indexers: Dict[str, TokenIndexer] = None,
+                 cross_validation_split_to_exclude: int = None,
+                 lazy: bool = False) -> None:
+        super().__init__(lazy)
+        self._token_indexers = token_indexers or {'tokens': SingleIdTokenIndexer()}
+        self._use_all_sql = use_all_sql
+        self._remove_unneeded_aliases = remove_unneeded_aliases
+        self._cross_validation_split_to_exclude = str(cross_validation_split_to_exclude)
+
+        self._sql_table_context = Text2SqlTableContext(schema_path)
+        self._world = Text2SqlWorld(self._sql_table_context)
+
+    @overrides
+    def _read(self, file_path: str):
+        """
+        This dataset reader consumes the data from
+        https://github.com/jkkummerfeld/text2sql-data/tree/master/data
+        formatted using ``scripts/reformat_text2sql_data.py``.
+
+        Parameters
+        ----------
+        file_path : ``str``, required.
+            For this dataset reader, file_path can either be a path to a file `or` a
+            path to a directory containing json files. The reason for this is because
+            some of the text2sql datasets require cross validation, which means they are split
+            up into many small files, for which you only want to exclude one.
+        """
+        files = [p for p in glob.glob(file_path)
+                 if self._cross_validation_split_to_exclude not in os.path.basename(p)]
+
+        for path in files:
+            with open(cached_path(path), "r") as data_file:
+                data = json.load(data_file)
+
+            for sql_data in text2sql_utils.process_sql_data(data,
+                                                            use_all_sql=self._use_all_sql,
+                                                            remove_unneeded_aliases=self._remove_unneeded_aliases,
+                                                            # TODO(Mark): Horrible hack, remove
+                                                            schema=self._sql_table_context.schema):
+                instance = self.text_to_instance(sql_data.text, sql_data)
+                if instance is not None:
+                    yield instance
```

Comment:
You probably need some flag somewhere to keep validation instances that aren't parseable, so we don't get artificially-inflated metrics.

Reply:
Yep - I will do that later, it should be easy to modify. We cover enough of the datasets completely that I can use one which we do cover 100% of initially. Thanks.